proposals/interface-types/working-notes/Instructions.md

```diff
@@ -0,0 +1,369 @@
+
+## `bool.lift_iMM`
+
+This instruction has type `[iMM] -> [bool]`.
+
+The output is `true` iff the input `iMM` has any bit set in the integer.
+
+## `bool.lower_iMM`
+
+This instruction has type `[bool] -> iMM`.
+
+The output is `0b1` zero extended to the desired bitwidth iff the input is true.
+
+## `sNN.lift_iMM`
+
+This instruction has type `iMM` -> `sNN`.
+
+The output is the value of `iMM` as a two's complement signed integer.
+
+If this cannot be represented in the desired type, then this instruction traps.
+
+## `sNN.lower_iMM`
+
+This instruction has type `[sNN] -> [iMM]`
+
+The output is the value of `sNN` as a two's complement signed integer.
+
+If this cannot be represented in the desired type, then this instruction traps.
+
+## `uNN.lift_iMM`
+
+This instruction has type `iMM` -> `uNN`.
+
+The output is the value of `iMM` as an unsigned integer.
+
+If this cannot be represented in the desired type, then this instruction traps.
+
+## `uNN.lower_iMM`
+
+This instruction has type `[uNN] -> [iMM]`
+
+The output is the value of `uNN` as an unsigned integer.
+
+If this cannot be represented in the desired type, then this instruction traps.
+
+## `string.lift_memory`
+
+The full syntax of this instruction is:
+```
+string.lift_memory #memidx #encoding
+```
+
+The type of this instruction is:
+```
+[$base: i32, $len: i32] -> [string]
+```
+
+The output is a sequence of unicode code points taken by interpreting `store.memory[$memidx].bytes[$base .. $base + $len]` with `#encoding`.
+
+This instructions traps if the range `$base .. $base + $len` is out-of-bounds, or the byte sequence is not valid under `#encoding`.
+
+The syntax for supported encodings is:
+```
+#encoding ::=
+  | utf8
+  | utf16
+```
+
+## `string.lower_memory`
+
+The full syntax of this instruction is:
+```
+string.lower_memory #memidx #encoding $alloc: #funcidx
+```
+
+The type of this instruction is:
+```
+[string] -> [i32, i32]
+
+Where:
+  * $alloc has a valid functype of [i32] -> [i32]
+```
+
+The instruction will execute the following steps:
+```
+  1. Let $length be the bytes required to encode the string with #encoding
+  2. Call $alloc with $length and receive an offset into #memidx
+  3. Encode the string into the offset at memory
+```
+
+## `record.lift`
+
+The full syntax of this instruction is:
+```
+record.lift #typeidx
+```
+
+The type of this instruction is:
+```
+[$a, $b, ..] -> [$type]
+
+Where:
+  * $type is a type of the form (record $fields) taken from #typeidx
+  * $a, $b, .. are the #interface-valtype's for each field in $fields
+```
+
+The output is a record interface value with fields given by the inputs.
+
+## `record.lower`
+
+The full syntax of this instruction is:
+```
+record.lower #typeidx
+```
+
+The type of this instruction is:
+```
+[$type] -> [$fields]
+
+Where:
+  * $type is a valid type of the form (record $fields) taken from #typeidx
+```
+
+The outputs are the fields of the input `record`.
+
+## `variant.lift #typeidx $tag: #u32`
```

Comment:
So I think what we want is for variant lifting/lowering to have a block control-flow structure with one sub-block for each case:

Lifting would have *roughly* the form:
```wasm
variant.lift #blocktype
  condition #caseidx
    ;; wasm expression evaluated with #blocktype params resulting in an i32 bool indicating this case was a match
  body
    ;; adapter code consuming  #blocktype params and producing an interface value of #caseidx's interface type
  end
  case #caseindx
    ...
  end
end
``` 
Where the conditions are evaluated in linear order to determine which case to lift into.  Lowering would have the simplified form.
```wasm
variant.lower #blocktype
  condition #caseidx
    ;; adapter code to lower from the #caseidx interface type into #blocktype's core wasm result
  end
  condition #caseidx
    ...
  end
end
```
By folding the control flow into the lifting/lowering, when you fuse a `variant.lift` with a `variant.lower` (b/c laziness), you can neatly fuse all the lifting/lowering of each case, leaving you with just plain branching core wasm.

Reply:
In general, lifting to a variant will require a predicate. (Think about lifting from Null pointers to the optional type). In fact, all you need for lifting to a variant is a 'vary #case' instruction, that takes an arbitrary number of elements from the stack and returns a variant of the type.

Lowering does seem to require a case-like form. Each arm of the case would have on the stack the elements associated with a particular variant
E.g.,
either[a,b] ::= either(a) | other(b)

switch T          ; has a value of type either[a,b] on the stack
  case $either ; has an "a" on the stack
  end
  case $other  ; has a "b" on the stack
  end
end

This is symmetric with how records are handled with pack and unpack.

Reply:
If I understand correctly, it seems like the difference here is whether control flow is structured within lifting/lowering instructions or whether we can utilize core wasm control flow. Luke has the predicates embedded within the lift instruction, and Francis and I had control flow outside of the lift instruction allowing the lift instruction to take the tag as an immediate.

I understand the benefit of having control flow in the lifting instructions to allow for easy fusing, so I would lean towards Luke's approach. Regardless, this is a larger discussion so I've temporarily dropped the variant type to unblock this PR.